ipk-client.1
.TH IPK-CLIENT 1
.SH NAME
ipk-client \- client for computing mathematic tasks
.SH SYNOPSIS
.B ipk-client
.IR hostname
.SH DESCRIPTION
.B ipk-client
connects to a server, receives mathematic task, solves it and sends back the result.
.PP
The
.IR hostname
can be valid IPv4 or IPv6 address. Even a domain name is supported.
.PP
The communication between the client and the server is done by sending messages HELLO, SOLVE, RESULT and BYE.
.PP
The client begins the communication by setting up a TCP connection with the server on the port \fI55555\fR.
After that, the client sends the message HELLO \fIid\fR\\n, where \fIid\fR is a md5 hash of the student login.
.PP
The server can answer by the message SOLVE \fInumber\fR \fIoperation\fR \fInumber\fR\\n.
Allowed operations are +, -, *, /.
The client computes the mathematic task and sends the solution in the message RESULT \fIsolution\fR\\n.
If the client is unable to solve the mathematic task (dividing by zero, overflow), it sends the message RESULT ERROR\\n.
.PP
The server can continue by another SOLVE message, or by the message BYE \fIsecret\fR\\n.
The client writes out the value of \fIsecret\fR to the stdout and closes the connection.
If the secret is in format UNKNOWN, it means, that the server was unable to identify the entered login sent in HELLO messsage.
.SH RETURN VALUE
On success, the program returns EXIT_SUCCESS. If an error occurs, the program returns EXIT_FAILURE.
.SH EXAMPLES
Connecting to a public server:
.PP
.nf
.RS
ipk-client 147.229.13.210
ipk-client 2001:67c:1220:80c::93e5:dd2
ipk-client pcgregr.fit.vutbr.cz
.RE
.fi
.PP
Connecting to the localhost:
.PP
.nf
.RS
ipk-client 127.0.0.1
ipk-client ::1
ipk-client localhost
.RE
.fi
.PP
.SH BUGS
If the server unexpectedly closes the connection, the client will be still waiting for the BYE message ad infinitum.
.PP
At the same time, the client can respond only to one valid message from the server, even if it receives more than one at once.
.SH AUTHOR
Written by Marek Jankech.

ipk-client.cpp
/**
 * File:        ipk-client.cpp
 * Author:      Marek Jankech
 * Date:        23.4.2017
 * Project:     Client for the computation of mathematic operations
 * Description: The client connects to the server, receives a mathematic task,
 *              solves it and sends the solution to the server
 */

#include <cstdlib>
#include <cstdio>
#include <iostream>
#include <sstream>
#include <algorithm>
#include <unistd.h>
#include <string>
#include <cstring>
#include <climits>
#include <regex.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <stdbool.h>
#include "md5.h"

#define BUFSIZE 1024
#define DEC 10

using namespace std;

/**
 *Global constants
 */
const string login = "xjanke01";
const char *serverPort = "55555";
const char *solveRegex = "\n? *SOLVE ([+-]?[0-9]+ [+-\\*\\/] [+-]?[0-9]+)\n";
const char *byeRegex = "\n? *BYE ([^\n]*)\n";

/**
 *Prototypes of functions
 */
bool match(const char *pattern, char *haystack, string &matched);
vector<string> tokenize(string str);
void errTerminate(string msg);
int evaluate(vector<string> arr, long double *result);
string lDouble2Str(long double number);

int main(int argc, char *argv[])
{
  //Check the number of arguments
  if (argc != 2 ) {
    errTerminate("arguments");
  }

  string loginHash = md5(login);
  string msg = "";
  string exercise = "";
  string secret = "";
  vector<string> arr;
  char buffer[BUFSIZE];
  struct addrinfo hints;
  struct addrinfo *result;
  int clientSocket, ecode, bytesrx;
  long double solution;

  //Reset data structures
  memset(&hints, 0, sizeof(struct addrinfo));
  memset(buffer, 0, sizeof(buffer));
  //Specify criteria
  hints.ai_flags = 0;
  hints.ai_family = AF_UNSPEC;
  hints.ai_socktype = SOCK_STREAM;
  hints.ai_protocol = 0;
  //Address translation for socket
  if ((ecode = (getaddrinfo(argv[1], serverPort, &hints, &result))) != 0) {
    errTerminate(gai_strerror(ecode));
  }
  //Create socket
  if ((clientSocket = socket(result->ai_family, result->ai_socktype, result->ai_protocol)) == -1) {
		errTerminate("socket");
	}
  //Establish a connection to the server
  if (connect(clientSocket, result->ai_addr, result->ai_addrlen) == -1) {
    errTerminate("connect");
  }
  //Free addrinfo structure
  freeaddrinfo(result);
  //Send HELLO message
  msg = "HELLO " + loginHash + "\n";
  if ((send(clientSocket, msg.c_str(), msg.size(), 0)) == -1) {
    errTerminate("send");
  }
  //Communicate with the server
  while(1) {
    bytesrx = recv(clientSocket, buffer, BUFSIZE, 0);
    if (bytesrx == -1) {
      errTerminate("recv");
    }
    else {
      //Terminate string with '\0'
      buffer[bytesrx] = '\0';
      //BYE message from the server - print the secret and close connection
      if (match(byeRegex, buffer, secret)) {
        cout << secret;
        break;
      }
      //SOLVE message from the server - answer it
      else if (match(solveRegex, buffer, exercise)) {
        arr = tokenize(exercise);
        //The client cannot solve the task - sending ERROR
        if ((evaluate(arr, &solution)) == -1) {
          msg = "RESULT ERROR\n";
        }
        //The client can solve the task - sending solution
        else {
          msg = "RESULT " + lDouble2Str(solution) + "\n";
        }
        //cout << exercise << "\n";
        //cout << msg << "\n";
        //Sending response to the server
        if ((send(clientSocket, msg.c_str(), msg.size(), 0)) == -1) {
          close(clientSocket);
          errTerminate("send");
        }
      }
    }
  }
  close(clientSocket);
  return EXIT_SUCCESS;
}

/**
 * Function for terminating the program with an error message
 */
void errTerminate(string msg) {
  cerr << "Error: " << msg << "\n";
  exit(EXIT_FAILURE);
}

/**
 * Function for matching a pattern in the string
 */
bool match(const char *pattern, char *haystack, string &matched) {
  bool match = false;
  const int nmatch = 2;
  regmatch_t pmatch[nmatch];
  regex_t regexp;
  int status;
  char buf[BUFSIZE];

  //Compile regex
  if ((status = regcomp(&regexp, pattern, REG_EXTENDED)) != 0) {
    regerror(status, &regexp, buf, BUFSIZE);
    printf("error = %s\n", buf);
    errTerminate("regcomp");
  }
  //Execute the regex for the string
  if ((regexec(&regexp, haystack, nmatch, pmatch, 0)) == 0) {
    match = true;
    matched = ((string)haystack).substr(pmatch[1].rm_so, pmatch[1].rm_eo - pmatch[1].rm_so);

  }
  regfree(&regexp);
  return match;
}

/**
 * Function for tokenizing string
 */
vector<string> tokenize(string str) {
  vector<string> arr;
  stringstream ss(str);
  string tmp;

  while (ss >> tmp) {
    arr.push_back(tmp);
  }
  return arr;
}

/**
 * Function for evaluating an expression in vector
 */
int evaluate(vector<string> arr, long double *result) {
  const int denominator = 100;
  long long int operand1;
  long long int operand2;
  stringstream ss;
  char operation;
  long long int tmp;

  //Convert strings to values
  operand1 = strtoll(arr[0].c_str(), NULL, DEC);
  operand2 = strtoll(arr[2].c_str(), NULL, DEC);
  ss << arr[1];
  ss >> operation;
  //Test limits
  if (operand1 == LLONG_MAX || operand1 == LLONG_MIN || operand2 == LLONG_MAX || operand2 == LLONG_MIN) {
    return -1;
  }
  //Evaluate
  switch (operation) {
    case '+':
      *result = operand1 + operand2;
      break;
    case '-':
      *result = operand1 - operand2;
      break;
    case '*':
      *result = operand1 * operand2;
      break;
    case '/':
      //Test dividing by zero
      if (operand2 == 0) {
        return -1;
      }
      *result = operand1 / (long double)operand2;
      break;
  }
  //truncate result
  tmp = *result * denominator;
  *result = tmp / double(denominator);
  return 0;
}

/**
 * Function for converting long double to string with 2 two decimal spaces
 */
string lDouble2Str(long double number) {
  ostringstream ostrs;
  ostrs.precision(2);
  ostrs << fixed << number;
  return ostrs.str();
}
